### PR TITLE
increase coverage of building with LDC by using it for the x64-debug build

### PR DESCRIPTION
**Pull request by rainers**
also switches to LDC 1.19.0

in support of https://github.com/dlang/installer/pull/425

### CODE REVIEW

**Review by thewilsonator — APPROVED** [review]

**Comment by rainers** [comment]
Error code 5 is probably the low byte of 0xc0000005, i.e. the compiler crashes with an access violation. It doesn't appear with the same file in different runs, but the respective builds seem close together (fail11746.d in the latest run, fail_pretty_errors.d in a previous one: https://dev.azure.com/dlanguage/dmd/_build/results?buildId=7564&view=results). Maybe it has to do with other parallel builds? @kinke Did you come across something similar?

**Comment by kinke** [comment]
Hmm, no, not really. What's up with all the extra empty output lines?
```
 ... fail_compilation\fail_pretty_errors.d -verrors=context   ()
Test fail_compilation\fail_pretty_errors.d failed.  The logged output:
D:\a\1\s\generated\Windows\Debug\x64\dmd.exe -conf= -m64 -Ifail_compilation -verrors=0 -verrors=context   -L/OPT:NOICF  -odtest_results\fail_compilation -oftest_results\fail_compilation\fail_pretty_errors_0.obj  -c fail_compilation\fail_pretty_errors.d  -map nul.map


fail_compilation\fail_pretty_errors.d(20): Error: undefined identifier `a`



    a = 1;




    ^



fail_compilation\fail_pretty_errors.d-mixin-25(25): Error: undefined identifier `b`



fail_compilation\fail_pretty_errors.d(30): Error: cannot implicitly convert expression `5` of type `int` to `string`



    string x = 5;




               ^



fail_compilation\fail_pretty_errors.d(35): Error: mixin `fail_pretty_errors.testMixin2.mixinTemplate!()` error instantiating



    mixin mixinTemplate;




    ^







==============================

Test fail_compilation\fail_pretty_errors.d failed: expected rc == 1, but exited with rc == 5
```

**Comment by rainers** [comment]
> What's up with all the extra empty output lines?

I think there are too many \n -> \r\n conversions happening when printing the output. That should not be the problem here, as \r is removed when comparing outputs. The problem is "rc == 5".

**Comment by kinke** [comment]
> That should not be the problem here

I'm rather worried about the varying number of empty lines inbetween actual lines. I can't help thinking that there's an output issue here (=> only fail_compilation tests with output fail), and that this is connected to https://forum.dlang.org/thread/ditydoqigcdjrjpvaltu@forum.dlang.org, where the guy is seeing incorrect output and compiler crashes with `-v` and DMD on Windows too.

**Comment by rainers** [comment]
I managed to reproduce the failure locally with `make -j8`, but not when using run.d to execute the tests. It seems to happen pretty randomly with any of the fail_compilation tests.

I tried to catch the exception by adding a C++ wrapper with `__try/__except` to stop the crashing program with a message box to allow attaching to it, but that didn't catch any error.

Looking at the trace of `procmon`, I noticed the GC scan threads are started due to some allocations not going through `rmem` happening during initialization. When compilation fails, the compiler calls `exit`, but that skips proper termination of the D runtime. Disabling GC collections seems to help!

**Comment by kinke** [comment]
Interesting finding. See [here](https://github.com/dlang/dmd/blob/f2e129463cbc68253daed661f2ff2b307aea8a51/src/dmd/mars.d#L908-L918) for the reason GC disabling is currently deferred to `_Dmain` (i.e., after all module ctors).

**Comment by kinke** [comment]
So I guess it'd be best to disable GC parallelization for the compiler altogether, so that `-lowmem` plays nice with `exit` upon errors?

**Comment by rainers** [comment]
> Interesting finding. See [here](https://github.com/dlang/dmd/blob/f2e129463cbc68253daed661f2ff2b307aea8a51/src/dmd/mars.d#L908-L918) for the reason GC disabling is currently deferred to `_Dmain` (i.e., after all module ctors).

That's too late because the GC collection is triggered during module ctors:  https://github.com/dlang/dmd/blob/f2e129463cbc68253daed661f2ff2b307aea8a51/src/dmd/imphint.d#L34

Recent druntimes have easy access to the gc-config, setting `core.gc.config.config.disable = true;` should be simpler.

**Comment by rainers** [comment]
> So I guess it'd be best to disable GC parallelization for the compiler altogether, so that `-lowmem` plays nice with `exit` upon errors?

Maybe, but faster collection times are very useful to dmd because of its large memory consumption. The problem might appear in other programs, too, so it would be nice to understand it better. Maybe there is an issue when `exit` is called while the GC threads are still starting up (shutdown happens about 50 ms after thread creation, and the system might be too busy with parallel jobs to actually start these threads).

**Comment by rainers** [comment]
> Recent druntimes have easy access to the gc-config, setting `core.gc.config.config.disable = true;` should be simpler.

Too recent for most used host compilers, now modifying rt_options.

**Review by WalterBright — APPROVED** [review]

**Comment by kinke** [comment]
With that neat `rt_options` trick (didn't think of conditionally setting that one, but then again it most likely won't work for LDC ltsmaster...), the old code in `_Dmain` should be removed.

**Comment by rainers** [comment]
> the old code in `_Dmain` should be removed.

Yeah, thought so, too. If the other GC allocation calls are eliminated, it would be the GC.disable() call that instantiates the GC.

**Comment by kinke** [comment]
> That's too late because the GC collection is triggered during module ctors

Thx, that also explains some other weird segfaults I've sporadically seen. I've fixed the symptoms with https://github.com/ldc-developers/ldc/pull/3306/commits/ee371b444730e3cf5b409a015b99c5b096328d7e but couldn't explain why stuff was being garbage collected...

**Review by kinke — APPROVED** [review]

File: src/dmd/root/rmem.d [hunk 1]
```diff
@@ -147,6 +148,8 @@ extern (C++) struct Mem
 
         static void disableGC() nothrow @nogc
         {
+            __gshared string[] disable_options = [ "gcopt=disable:1" ];
+            rt_options = disable_options;
```
Comment (MoonlightSentinel) [inline src/dmd/root/rmem.d]: Should this discard all other options?

Comment (rainers) [reply src/dmd/root/rmem.d]: It only replaces the currently embedded empty options, command line options still have higher precedence.
If the frontend is used as a library, you might want to use different options (though probably with the GC enabled anyway), so maybe it should be changed outside of rmem.d, e.g. mars.d?

File: src/dmd/mars.d [hunk 2]
```diff
@@ -902,21 +902,8 @@ version (NoMain) {} else
...
-        // Older host druntime versions need druntime to be initialized before
-        // disabling the GC, so we cannot disable it in C main above.
-        static if (isGCAvailable)
-        {
-            if (!mem.isGCEnabled)
-                GC.disable();
-        }
-        else
-        {
-            GC.disable();
```
Comment (kinke) [inline src/dmd/mars.d]: This is skipped now, in the unlikely event of `!isGCAvailable`.

Comment (rainers) [reply src/dmd/mars.d]: Good catch! I restored that part.

Comment (kinke) [reply src/dmd/mars.d]: I think you'll still need the `core.memory` import.

Comment (rainers) [reply src/dmd/mars.d]: > I think you'll still need the `core.memory` import.

Oops, restored. I also now moved setting rt_options to mars.d, where you would usually deal with these options to begin with.

**Comment by thewilsonator** [comment]
Is this good to go?

**Comment by rainers** [comment]
> Is this good to go?

Yes, I think so

**Pull request merged**